### PR TITLE
Documentation; Close to final APIs; Working async streams and pipelines

### PR DESCRIPTION
**Pull request by zah**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,7 +6,498 @@
...
+        error(UnexpectedEndOfFile)
+    of '"'
+      return
+    else:
+      result.add nextChar
+
+  error(UnexpectedEndOfFile)
+
+proc nextToken(stream: InputStream): JsonToken {.fsMultiSync.} =
+  while stream.readable:
```
Comment (arnetheduck) [inline README.md:217]: if this is the only async point, why not make it `while await stream.readable`? for sync streams, this would be a noop, for async streams it would do the future dance (if necessary).

it remains the case that nim has no safeguards against concurrent modification - teaching developers yet another hidden out-of-order execution keyword is a high cost in terms of bugs - in terms of syntactic overhead, this is almost nothing, but reminds the developer that _this_ is where the synchronization happens, and not in the `read` below, in a way that's established across libraries and frameworks, nim and not

File: README.md [hunk 2]
```diff
@@ -6,7 +6,498 @@
...
+If the output fits within a single page, it will be efficiently moved to
+the `getOutput` result. When the output size is known upfront you can ensure
+that this optimization is used by calling `stream.ensureRunway` before any
+writes, but please note that the library is free to ignore this hint in async
+context or if a maximum memory usage policy is specified.
+
+In a non-memory stream, any writes larger than a page or issued through the
+`writeNow` API will be sent to the output device immediately.
+
+Please note that even in async context, `write` will complete immediately.
```
Comment (arnetheduck) [inline README.md:437]: I wonder about this default: ie incomplete writes for files is such a common mistake and it's defaults like this in the API that cause it - on networks, it's even more common that writes fail. `tryWrite` perhaps would remind people of this fact, because I'd agree in general that a function that doesn't wait/block often is a good tradeoff

Comment (arnetheduck) [reply README.md:437]: basically, this is where data corruption happens, causing users to lose their work because developers don't consider the hidden delayed action - it's satisfying to the developer that feels that the library is easy to use because they don't have to bother with the details, but less satisfying to the user

Comment (arnetheduck) [reply README.md:437]: furthering this point - `when await s.readable` is a natural and "forced" collection point where error checking is done for reading the stream whereas for writes it remains "optional" - what would be a way to introduce a similar "forced" checkpoint for writes? if it's not "naturally" part of the API, it won't get called (just like `close` de-facto doesn't get caled with users paying the price for the poor API)

File: README.md (hunk 1 above)
Comment (arnetheduck) [reply README.md:217]: notably, `readable` does not necessarily have to be future-based - the new `await` (perhaps better call it `fsawait` to not overload `await` until it's clear from common usage that it's fine) could be a noop for sync streams and not require `Future` return types from whatever it calls - if it returns a future, it would do the future dance, if it returns `Opt[future]` or some such it would do the "do I need to alloc-a-future"-thing etc

File: README.md [hunk 3]
```diff
@@ -6,7 +6,498 @@
...
+
+  For reading memory mapped files which provides best performance.
+
+* `unsafeMemoryInput`
+
+  For handling strings, sequences and openarrays as an input stream. <br />
+  You are responsible for ensuring that the backing buffer won't be invalidated
+  while the stream is being used.
+
+* `memoryInput`
```
Comment (arnetheduck) [inline README.md:145]: put this one above `unsafeMemoryInput`? it's simpler to use and a good default in nim's GC world

File: README.md (hunk 2 above)
Comment (arnetheduck) [reply README.md:437]: another point: ignoring backpressure and making it "optional" will not fly - it's not an aspect that you _can_ ever ignore - you have to account for it in every application that does async writing or your application is broken _in the cases where performance actually matters, ie when writing a lot of data_, so there has to be a point in the faststreams API that forces the user to deal with it. `write` is the traditional place in `unix` etc - either it's synchronous or it returns `EAGAIN` - if faststreams is to offer an async mode, it necessarily needs to have a forced synchronization point that the caller of the code cannot ignore.

it really doesn't matter that the library is fast when writing 1 byte if it unexpectedly breaks with OOM when it writes 1gb.

Comment (arnetheduck) [reply README.md:437]: practical example: the USB stick that you have to "eject" before removing safely - in the 00's this was thought of as a great trade off  because it was _so fast_ but the world leaned better and now doesn't force users to go through this silliness - instead, copy operations visibly don't terminate until the writing is physically done.

File: README.md (hunk 1 above)
Comment (zah) [reply README.md:217]: It's possible to implement this and I might agree with it if it's what it takes to get wider adoption of the API, but I consider the argument a bit moot, because in the README I've highlighted that the FastStreams API is mostly useful for implementing:


* Parsers for data formats and protocols employing formal grammars
* Block ciphers
* Compressors and decompressors
* Stream multiplexers

Think about your concerns in this context. What is the mutable state in a block compressor or a Json parser? These are stream processing algorithms that are very much focused on their inputs and outputs and that benefit significantly from the "just like sync code" mentally of the FastStreams API.

File: README.md (hunk 2 above)
Comment (zah) [reply README.md:468]: I consider this a bit experimental myself and might agree to change the default. The point is that the library offers:

* `write` (buffers data and puts a future in `lastWriteFut` when you can check it if you wish) 
* `writeAndWait` (the standard write with back-pressure)
* `writeNow` (this a non-buffered write) 

It's just a matter of selecting which of the two should be promoted as the default way. If we make `writeAndWait` the default, what would be a good name for `write`?

File: README.md (hunk 3 above)
Comment (zah) [reply README.md:145]: But it's so must more inefficient! oh, well :)

File: README.md (hunk 1 above)
Comment (arnetheduck) [reply README.md:217]: well - the output goes somewhere - and as soon as you get into use cases where performance actually matters, as opposed to toy examples and config files, people will want to optimize where that is - for example a state variable (say.. a beacon state) that might get used or mutated from outside because nim doesn't track ownership. 

This means in particular, that for an extended period of time, partially updated instances will live in memory and be accessible "outside" - clever optimizations like RVO even mutate instances in a totally unexpected way - even the most experienced compiler developers get this wrong though they intimately know the feature - so you can't even blame lazy devs that didn't read the manual (https://github.com/nim-lang/Nim/issues/14126).

On a network (where async is predominatly used), data trickles slowly, with latency, errors, disconnects and all kinds of erratic behaviours - to write something that approaches correct and performant behaviour, you need to adapt your code to where the reading and waiting happens, and for that, you need to know and see it. A readme in a far-off repository doesn't count, because that readme is not on your mind when you're trying to solve a problem - you're busy with something else.

In short, you can't responsibly hide this stuff, it only works when it doesn't matter - on the toy examples that devs will try and call "testing".

Comment (zah) [reply README.md:217]: The output will usually go at the end of a "Pipeline". Here is one example:

https://github.com/status-im/nim-beacon-chain/blob/b41f4347731585f66630a1a02e3f15326acaf4a8/beacon_chain/faststreams_backend.nim#L90-L91

Such pipelines are usually awaited.

It also occurred to me that there is not much evidence that a visible `await` prevents the problem. Writing `await` is pretty routine and people don't usually stop to think about these issues. I'll agree that someone who is extremely knowledgable about async issues, but not familiar with FastStreams may benefit from a visible `fsAwait` during code review (well, she'll have to know what that is).

Comment (arnetheduck) [reply README.md:217]: anybody who is not knowledgeable about `async` issues will contribute broken code to the world with this API - it's inviting the developer to not understand the issues while not doing the right things to solve them.

writing `await` gives whoever is reading the code a hint about what might be wrong, and where - much like a function signature, it's a marker that shows where specific issues need to be considered - this is helpful not only during development but also later, during debugging. it also opens up the mind to the right google terms to search for, which gives access to a wide range of resources across languages that explain these issues.
